### ISSUE TITLE
[TEST] Add stress testing for fs volume

### ISSUE THREAD

**Issue by innobead** [longhorn/longhorn#6379]
## What's the test to develop? Please describe

Could consider using https://github.com/ColinIanKing/stress-ng or fio to stress test Longhorn fs volumes to see if the volume still runs well w/o issues like data corruption or data integrity. 

Discuss with @derekbit and @shuo-wu when working on this. 

## Describe the items of the test development (DoD, definition of done) you'd like
> Please use a task list for items on a separate line with a clickable checkbox https://docs.github.com/en/issues/tracking-your-work-with-issues/about-task-lists

- [ ] `item 1`

## Additional context

Add any other context or screenshots about the test request here.

**Comment by innobead** [comment]
@yangchiu you will be the mentor for @nitendra-suse on this test.

**Comment by yangchiu** [comment]
We have a rough test plan about disk stress on the negative testing [confluence page](https://confluence.suse.com/pages/viewpage.action?spaceKey=LON&title=Negative+Testing). Like:

- Make file system on node as RO
  -  While the operations are running, make the file system on the nodes as RO
- Disk Full	
  - Almost fill up the Longhorn disk like 90%. Check the Longhorn behavior.
- Disk with low IOPS	
  - Add an HDD disk on Longhorn volume with low IOPS or use some tool to decrease the IOPS
- Certain blocks failure	
  - Pause and make some block on the device corrupt? Make some sectors on disk bad?
- Hard drive SMART detected failure	
  -  Introduce some SMART failure?
- Disk unscheduling	
  - While the operations are running, try to unschedule the disks using Longhorn UI or by ssh to nodes.

But this test plan is still in an early stage and without concrete ideas about how to practice them, and we also don't know if they're good enough to test and guarantee what we'd like to ensure.

It would be appreciated if @derekbit and @shuo-wu  can help to provide any information to correct anything missing, improve the coverage, or even the practical ways to execute these tests. 

And also @nitendra-suse feels free to update anything here and make progress together. Many thanks!

**Comment by yangchiu** [comment]
Have discussed with @derekbit. 

The goal this issue wants to achieve is quite straightforward: 

**- Have a long running (>12 hrs) pipeline keeps writing/reading data through/from Longhorn volume (both ext4 and xfs type should be tested), and see if there's a very small chance that data will corrupt.**

It should create a pod using Longhorn volume (like what we frequently did in longhorn automated test cases). And in the pod, using `fio` or other tools like `stress-ng` to keep writing/reading. So the path would be something like:
```
pod (fio) --rw--> longhorn volume (xfs, ext4)
```

We can try `fio` first. It seems has the ability to automatically verify data when reading data back (using the `--do_verify` option), which is not provided in `stress-ng`.

When using `fio`, remember to:
- mix read/write operations instead of 100% read or write, to simulate real world scenario
- use multithreads (workers > 1)
- iodepth = 64 or 32
- ioengine=libaio

Reference https://fio.readthedocs.io/en/latest/fio_doc.html for more information.

cc @nitendra-suse @longhorn/qa @innobead

**Comment by shuo-wu** [comment]
Will we add any interventions like workload crash or network disconnection during the test? Keeping IO without any other operations may not be able to trigger fs corruption.

**Comment by yangchiu** [comment]
> Will we add any interventions like workload crash or network disconnection during the test? Keeping IO without any other operations may not be able to trigger fs corruption.

Sounds good to me. WDYT @derekbit ?

@shuo-wu @derekbit It would be appreciated if you can help to provide any information to complete the test scenarios, improve the coverage, or even the practical ways to execute these test steps.

@nitendra-suse also feel free to update anything here, like any thoughts, any progress or any discussion you've had.

**Comment by shuo-wu** [comment]
IMO, the test can be like:
1. Launch a pod that keeps writing and reading data to/from a Longhorn volume.
2. Wait for a random time(I think the min time can be >= 1hour), directly crashing (or forcibly killing without wait time) the pod.
3. Restart the pod with the old volume and continue the IO
4. Repeat step 1 ~ 3

**Comment by nitendra-suse** [comment]
So far came up with fio command to use : 
fio --name=global --rw=randrw --filename=/var/lib/longhorn/ --direct=1 --ioengine=libaio --iodepth=32 --bs=1M --runtime=3600 --numjobs=16 --verify=sha1 --do_verify=1

**Comment by derekbit** [comment]
> > Will we add any interventions like workload crash or network disconnection during the test? Keeping IO without any other operations may not be able to trigger fs corruption.
> 
> Sounds good to me. WDYT @derekbit ?

We can leverage the chaos engineering https://blog.mayadata.io/openebs/storage-chaos-engineering-with-litmus-an-overview-1. 

> 
> @shuo-wu @derekbit It would be appreciated if you can help to provide any information to complete the test scenarios, improve the coverage, or even the practical ways to execute these test steps.
> 
> @nitendra-suse also feel free to update anything here, like any thoughts, any progress or any discussion you've had.

**Comment by yangchiu** [comment]
> 
> For a more complicated or systematic perspective, we probably can leverage the chaos engineering https://blog.mayadata.io/openebs/storage-chaos-engineering-with-litmus-an-overview-1.
> 

Yes, [Litmus](https://docs.litmuschaos.io/) has been mentioned several times in our negative testing discussion. But so far no experiment has been done to test if it's possible to integrate it into our new testing framework. Probably we can give it a try since our new negative testing framework is still in early stage and it's easier to adopt new things. cc @longhorn/qa @innobead @weizhe0422

**Comment by yangchiu** [comment]
> So far came up with fio command to use : 
> 
> fio --name=global --rw=randrw --filename=/var/lib/longhorn/ --direct=1 --ioengine=libaio --iodepth=32 --bs=1M --runtime=3600 --numjobs=16 --verify=sha1 --do_verify=1

Do we have any further progress on this? Can we try to integrate this into a negative test step now?

**Comment by nitendra-suse** [comment]
> Do we have any further progress on this? Can we try to integrate this into a negative test step now?

I am trying to use 
`dd if=/dev/zero of=filename bs=$((1024*1024)) count=$((10*1024)) `
command by ssh into the node and run this one and try to fill say 80% or 90%.
Otherwise we can integrate this one
`fio --name=global --rw=randrw --filename=/var/lib/longhorn/ --direct=1 --ioengine=libaio --iodepth=32 --bs=1M --runtime=3600 --numjobs=16 --verify=sha1 --do_verify`=1
